### ISSUE TITLE
NAF results reproductibility

### ISSUE THREAD

**Issue by AmandineBtto** [facebookresearch/real-acoustic-fields#4]
Hello, 

I've tried to reproduce paper results of vanilla NAF on RAF. 

I kept NAF github code and followed your paper indications: 
- I've changed the grid and the poses to be 3D instead of 2D.
- I've cut the RIR to 0.32s and use the 48kHz sample rate with STFT parameters (Nfft 1024, win length 512, hop length 256).
- For the split I randomly sample 80% of each room for train, 15% for test and 5% for val. 
- I use the random phase waveform reconstruction.
- Optimizer is AdamW, beginning at lr = 1e-3 with a decay rate of 0.98.
- batch size of 128, training for 200 epochs.

However, when averaging on the two scenes I obtain: STFT error = 0.2347, T60 = 12.38% , C50 = 3.59 dB and EDT = 0.111 s.
Those results are quite different from to Table 2 results (STFT error 0.64, T60 10.08, C50 0.71,  EDT 0.021).

Could you provide a code example showing how you have adapted one of the presented methods to fit your dataset? 
Am I missing something? 

Thank you!

**Comment by IFICL** [comment]
Hi, 

For the NAF baseline, we based it on their official code and extended it to a 3D grid simply. I think the setup you have is correct. 

T60 and EDT errors are different might be because we measure early decay time in a second by approximate decay of 60dB using 10dB. When measuring T60 error, we run a high pass filter to filter out the frequency under 200Hz.

```
def measure_edt(h, fs=22050, decay_db=10):
    '''
        Measure early decay time in seconds: we approx decay of 60dB using 10dB
        h (signal): `(T, )` np.array
        fs: sampling rate
    '''
    fs = float(fs)

    # The power of the impulse response in dB
    power = h ** 2
    energy = np.cumsum(power[::-1])[::-1]  # Integration according to Schroeder

    # remove the possibly all zero tail
    i_nz = np.max(np.where(energy > 0)[0])
    energy = energy[:i_nz]
    energy_db = 10 * np.log10(energy)
    energy_db -= energy_db[0]

    i_decay = np.min(np.where(- decay_db - energy_db > 0)[0])
    t_decay = i_decay / fs
    # compute the decay time
    decay_time = t_decay
    est_edt = (60 / decay_db) * decay_time
    return est_edt

def measure_rt60_advance(signal, sr, decay_db=10, cutoff_freq=200):
    # apply highpass filter
    signal = ta.functional.highpass_biquad(
        waveform=signal,
        sample_rate=sr,
        cutoff_freq=cutoff_freq
    )
    signal = signal.cpu().numpy()
    rt60 = measure_rt60(signal, sr, decay_db=decay_db, plot=False)
    return rt60
```

Your STFT error is better than ours while C50 is worse. We are unsure about this part. We find NAF doesn't do good on STFT error. 

To provide your example code, we will need approval from Meta. I am afraid that we cannot offer them at this stage.

**Comment by IFICL** [comment]
Another thing we want to mention is that. To ensure a fair comparison between INRAS, the NAF's outputs are converted to waveform domain for evaluation. So for STFT error, we will use converted waveform and reperform the STFT operation for evaluation. This might introduce some noise to the NAF baseline since the final output we want is RIR in waveform

**Comment by AmandineBtto** [comment]
Thank you for all the information! 

For T60, I was computing the value for 30dB decay and then extrapolating it to 60dB. I used your code with the high-pass filter + decay of 10dB extrapolated to 60dB but I obtain even higher values : T60 = 28.55% (mean on both rooms). Also, it seems that using 10dB instead of 30dB changes the metric a lot. Do you have any insight on this? 

For EDT, I'm using the same code as you so I don't understand what causes the difference.

For C50, this is the code I'm using: 

```
def measure_clarity(signal, time=50, fs=48000):
    # INRAS
    h2 = signal**2
    t = int((time/1000)*fs + 1) 
    return 10*np.log10(np.sum(h2[:t])/np.sum(h2[t:]))
```
    
 Is it matching yours? 
 
I've also adopted your method for the STFT error: First, I go back from STFT to waveform.  Then I do log STFT again.  Finally I compute the error with the log STFT from the ground truth waveform. 
With this method I obtain an error of 0.34 dB.  It is still quite far away from the paper. 

How are you performing inverse STFT? I'm using the "towave" function of NAF (test_utils.py) with ISTFT parameters that match the paper's for 48kHz.

**Comment by IFICL** [comment]
For ISTFT, this is how we do 
```
def spec2wav(self, mag):
        # recover the spec
        with torch.no_grad():
            N, C, F, T = mag.shape
            mag = mag * self.hparams.stft.std.to(mag.device) + self.hparams.stft.mean.to(mag.device)
            mag = torch.exp(mag) - self.hparams.stft.offset

            mag = mag.contiguous().view(N * C, F, T)
            phase = (torch.rand_like(mag) - 0.5) * 2 * np.pi # put the random phase 
            spec = torch.polar(mag, phase)
            wav = torch.istft(
                input=spec,
                n_fft=self.hparams.stft.n_fft,
                hop_length=self.hparams.stft.hop_length,
                win_length=self.hparams.stft.win_length
            )
            wav = wav.view(N, C, wav.shape[-1])
        return wav
```

Your implementation of C50 looks good to me. We are unsure why you obtained a different performance from us. We significantly change the code of NAF with our preprocessing and network implementations. 

Perhaps you can start with INRAS to check if those performances are matched or not.

**Comment by AmandineBtto** [comment]
Hello, 

For NAF, I've adopted your previous code extracts and here are my metrics : STFT error = 0.419, T60 = 28.8%, C50 = 3.60, EDT = 0.113. 
As you suggested, I've checked the results on INRAS. I applied the following changes to the original code including:
- Poisson sampling on the mesh to obtain 256 bounce points
- going from 2D to 3D 
- STFT multiresolution loss. 

I also use your advanced T60 implementation. The results are STFT error = 0.262, T60 = 16.14%, C50 = 1.50, EDT = 0.0463. 
It's also quite different from the paper's results.
Do you have any idea about what may cause the difference? 

I was also wondering which angle you were using for the source orientation. I suppose you use the quaternion to obtain Euler angles but then are you giving the angle on each axis or only the one around the up axis? 

Thank you so much for you help!

**Comment by IFICL** [comment]
To get the euler angle:

```
r = R.from_quat(spk_rot)
view_dir = np.array([1, 0, 0])
spk_viewdir = r.apply(view_dir)

spk_rot = r.as_euler('yxz', degrees=True)
spk_rot = np.round(spk_rot[0], decimals=0)
spk_rot = torch.tensor([spk_rot / 180.0]).float() # normalize the euler angle to [-1, 1]
```

We are unsure why there is a large performance gap. For our implementation, especially INRAS, we have double-checked the authors and fixed the issue in their paper (by removing the unused phase loss). We verified our implementation on the SoundSpace 1.0 dataset and we are able to get a similar performance.

**Comment by ZitongLan** [comment]
Hi @IFICL can you elaborate more on the euler angle you are using? I get confused about the quaternion to euler angle transformation. From my understanding, euler angle should be a 3 value vector. But from your code snippet, it only returns one angle, is this angle suppse to be the orientation along the xz plane (or namely horizontal plane)

**Comment by IFICL** [comment]
@ZitongLan 
Hi, Yes. You are right. The speaker rotation only happens in the horizontal plane. To simplify the inputs, we only use a single angle. We also experiment with using quaternion directly, while we find the single Euler angle representation has more stable training.

**Comment by ZitongLan** [comment]
@IFICL Thanks for you quick resposne! I have a follow up question about the euler angle you calculated, since it is in the xz plane, I find that the value is in degree ranging from -180 to 180, say denoted as theta. Then is the direction represented in xz vector (cos(theta), sin(theta)) in xz plane? For example, if the angle I get is -90, does it mean it is oriented to the negative direction of z axis?

**Comment by IFICL** [comment]
@isrish Hi, Isreal. Could you answer the speaker's orientation question from @ZitongLan?

**Comment by isrish** [comment]
> @IFICL Thanks for you quick resposne! I have a follow up question about the euler angle you calculated, since it is in the xz plane, I find that the value is in degree ranging from -180 to 180, say denoted as theta. Then is the direction represented in xz vector (cos(theta), sin(theta)) in xz plane? For example, if the angle I get is -90, does it mean it is oriented to the negative direction of z axis?

A -ve 90-degree rotation is clockwise so that a vector originally aligned with the +ve x-axis becomes aligned with the +ve z-axis. A  +ve 90-degree rotation is counter-clockwise, it will align the speaker to the -ve z-axis. Note that the speaker only have rotations about the y-axis.

**Comment by swapb94** [comment]
@AmandineBtto, were you to reproduce the scores for NAF, INRAS?

**Comment by ChillCoing** [comment]
> @AmandineBtto, were you to reproduce the scores for NAF, INRAS?

Hi @swapb94 , what is your reproduced C50 and EDT result? It seems my C50 is about 0.9dB and EDT is about 23ms (can not achieve 0.5dB and 19ms)
